contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,182 @@
+     * call to the next layer's forwarder contract.
+     */
+    function relayMessage(address target, bytes memory message) external payable override onlyAdmin {
+        RouteIndex storage route = availableRoutes[target];
+        bool success;
+        // Case 1: We are on the network immediately before our target, so we forward the original message over the canonical messaging bridge.
+        if (target == route.spokePool || target == route.forwarder) {
+            (success, ) = route.adapter.delegatecall(abi.encodeCall(AdapterInterface.relayMessage, (target, message)));
+        } else {
+            // Case 2: We are not on the network immediately before our target, so we wrap the message in a `relayMessage` call and forward it
+            // to the next forwarder on the path to the target contract.
+            bytes memory wrappedMessage = abi.encodeCall(AdapterInterface.relayMessage, (target, message));
+            (success, ) = route.adapter.delegatecall(
+                abi.encodeCall(AdapterInterface.relayMessage, (route.forwarder, wrappedMessage))
+            );
+        }
+        if (!success) revert RelayMessageFailed();
+        emit MessageForwarded(target, message);
+    }
+
+    /**
+     * @notice Relays `amount` of a token to `target` over an arbitrary number of layers.
+     * @param baseToken This layer's address of the token to send.
+     * @param remoteToken The next layer's address of the token to send.
+     * @param amount The amount of the token to send.
+     * @param target The address of the contract that will receive the tokens.
+     * @dev The relayTokens function (which must be implemented on a per-chain basis) needs to inductively determine the address of the
+     * remote token given a known baseToken. Then, once the correct addresses are obtained, this function may be called to finish the
+     * relay of tokens.
+     */
+    function _relayTokens(
+        address baseToken,
+        address remoteToken,
+        uint256 amount,
+        address target
+    ) internal {
+        RouteIndex storage route = availableRoutes[target];
+        bool success;
+        // Case 1: We are immediately before the target spoke pool, so we send the tokens to the spoke pool
+        // and do NOT follow it up with a message.
+        if (target == route.spokePool) {
+            (success, ) = route.adapter.delegatecall(
+                abi.encodeCall(AdapterInterface.relayTokens, (baseToken, remoteToken, amount, target))
+            );
+            if (!success) revert RelayTokensFailed(baseToken);
+        } else {
+            // Case 2: We are not immediately before the target spoke pool, so we send the tokens to the next
+            // forwarder and accompany it with a message containing information about its intended destination.
+
+            // Send tokens to the forwarder.
+            (success, ) = route.adapter.delegatecall(
```

Comment:
Is this code chunk the same as in the Rerouter adapter?

Reply:
Yes exactly.